devnet-sdk/systemgo/l2_challenger.go

```diff
@@ -0,0 +1,124 @@
+package systemgo
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+
+	"github.com/ethereum-optimism/optimism/devnet-sdk/system2"
+	"github.com/ethereum-optimism/optimism/op-chain-ops/devkeys"
+	chconfig "github.com/ethereum-optimism/optimism/op-challenger/config"
+	"github.com/ethereum-optimism/optimism/op-challenger/game"
+	chmetrics "github.com/ethereum-optimism/optimism/op-challenger/metrics"
+	"github.com/ethereum-optimism/optimism/op-service/metrics"
+)
+
+type L2Challenger struct {
+	service *game.Service
+	// no challenger RPC endpoint, yet
+}
+
+func WithChallenger(challengerID system2.L2ChallengerID,
+	l1ELID system2.L1ELNodeID, l1CLID system2.L1CLNodeID,
+	l2CLID system2.L2CLNodeID, l2ELID system2.L2ELNodeID, supervisorID system2.SupervisorID) system2.Option {
+	return func(setup *system2.Setup) {
+		orch := setup.Orchestrator.(*Orchestrator)
+		setup.Require.False(orch.challengers.Has(challengerID), "challenger must not already exist")
+
+		challengerSecret, err := orch.keys.Secret(devkeys.ChallengerRole.Key(challengerID.ChainID.ToBig()))
+		setup.Require.NoError(err)
+
+		logger := setup.Log.New("id", challengerID)
+		logger.Info("Challenger key acquired", "addr", crypto.PubkeyToAddress(challengerSecret.PublicKey))
+
+		l2ID := setup.System.L2NetworkID(challengerID.ChainID)
+		l2Net := setup.System.L2Network(l2ID).(system2.ExtensibleL2Network)
+		disputeGameFactoryAddr := l2Net.Deployment().DisputeGameFactoryProxyAddr()
+
+		// TODO: refactor systemgo orchestrator to make endpoints available by ID
+		l1EL, ok := orch.l1ELs.Get(l1ELID)
+		setup.Require.True(ok)
+		l1CL, ok := orch.l1CLs.Get(l1CLID)
+		setup.Require.True(ok)
+		l2CL, ok := orch.l2CLs.Get(l2CLID)
+		setup.Require.True(ok)
+		l2EL, ok := orch.l2ELs.Get(l2ELID)
+		setup.Require.True(ok)
+		super, ok := orch.supervisors.Get(supervisorID)
+		setup.Require.True(ok)
+
+		supervisorEndpoint := super.userRPC
+		l1Endpoint := l1EL.userRPC
+		l1Beacon := l1CL.beaconHTTPAddr
+		rollupEndpoint := l2CL.rpc
+		l2ELEndpoint := l2EL.userRPC
+
+		chDir := orch.t.TempDir()
+
+		cfg := chconfig.NewConfig(disputeGameFactoryAddr, l1Endpoint, l1Beacon, rollupEndpoint, l2ELEndpoint, chDir)
+		cfg.Cannon.L2Custom = true
+		// The devnet can't set the absolute prestate output root because the contracts are deployed in L1 genesis
+		// before the L2 genesis is known.
+		cfg.AllowInvalidPrestate = true
+		cfg.TxMgrConfig.NumConfirmations = 1
+		cfg.TxMgrConfig.ReceiptQueryInterval = 1 * time.Second
+		if cfg.MaxConcurrency > 4 {
+			// Limit concurrency to something more reasonable when there are also multiple tests executing in parallel
+			cfg.MaxConcurrency = 4
+		}
+		cfg.MetricsConfig = metrics.CLIConfig{Enabled: false}
+
+		// TODO: need equivalent of challenger-helper applyCannonConfig func, but not attached to e2e sys.
+		// Also need to fix the rollup-config/chain-config json writers in there,
+		// to not overwrite the same file for every config (?)
```

Comment:
DataDir is always set to a temp dir in `NewChallengerConfig` so while it writes to a consistent filename, it should always have a unique directory and so not be overwriting previous configs.

Reply:
Specifically I am looking at https://github.com/ethereum-optimism/optimism/blob/3a9b0c0ddf908e56b5b4073c98b14343d9acb3e9/op-e2e/e2eutils/challenger/helper.go#L146

There's a for loop, and it appears to iterate over configs, and write each of them to the same `filepath.Join(c.Datadir, "l2-genesis.json")` constant. The datadir isn't changed while looping.

And then there's another for-loop that does the same with the rollup-config.

Reply:
ah right, yeah that's a screw up - we haven't been able to validate any of the interop support in these helpers so far because it was blocked on OPCM supporting deploying the new game type.

Reply:
https://github.com/ethereum-optimism/optimism/pull/15091 fixes this.